Iris 2.3 What's New (#3450)
SciTools/iris: docs/iris/src/whatsnew/2.3.rst

What's New in Iris 2.3.0

> **********************

Author: Thanks, missed that after I added ".0"

> :Release: 2.3

Reviewer: @trexfeathers Can we explicitly clarify this as `2.3.0`

Author: 👍

:Date: 2019-10-04

This document explains the new/changed features of Iris in version 2.3.0
(:doc:`View all changes <index>`.)

Iris 2.3.0 Features

> =================

Author: Thanks, missed that after I added ".0"

.. _showcase:

.. admonition:: Increased Support for CF 1.7

  Several changes have contributed to our increased support for CF 1.7:

  * climatological coordinate support
  * standard name modifier support

>   * name tokens enforcement in cell methods

Reviewer 2: I don't think we should mention this as it is a bugfix so isn't really suitable in this section

>   * name tokens enforcement in cell methods

Author: OK it's gone

>   * coordinate var_name, CF_CONVENTIONS_VERSION advanced to 1.7

Author: This wasn't how I intended, hence it read strangely. Have fixed to what I wanted originally

>   * coordinate var_name, CF_CONVENTIONS_VERSION advanced to 1.7

Author: I have also added links to internal anchors. Unfortunately anchors can't be placed in the middle of a bullet list so the links will only take you to the relevant _section_. See what you think

  You can read more about each of these below.

.. admonition:: Climatological Coordinate Support

  Iris can now load, store and save `NetCDF climatological coordinates
  <http://cfconventions.org/Data/cf-conventions/cf-conventions-1
  .7/cf-conventions.html#climatological-statistics>`_. Any cube time
  coordinate can be marked as a climatological time axis using the boolean
  property: ``climatological``. The climatological bounds are stored in the
  coordinate's ``bounds`` property.

  When an Iris climatological coordinate is saved in NetCDF, the NetCDF time

>   axis will be given a 'climatology' attribute, and the contents of the

Reviewer 2: ```suggestion
  coordinate variable will be given a 'climatology' attribute, and the contents of the
```
From CF conventions:

> a climatological time coordinate variable does not have a bounds attribute. Instead, it has a climatology attribute, which names a variable with dimensions (n,2), n being the dimension of the climatological time axis

So its the variable that has a `climatology` attribute

Author: Thanks, I appreciate the guidance. I've read and re-read that CF section but it's taking me a while to get my head around all the terminology

  coordinate's ``bounds`` property are written to a NetCDF boundary variable
  called 'climatology_bounds'.  These are in place of the more usual NetCDF
  'bounds' attribute and accompanying 'bnds' boundary variable. See below
  for an example:

>     .. code-block::

Reviewer 2: This errors when building:
`2.3.rst:41: WARNING: Error in "code-block" directive:`
And the resulting built page misses the code block completely. 

```suggestion
    .. code-block:: none
```

Author: Interesting that this worked on my local build without "none". Anyway I've made this change and the new formatting still looks good 👍

Reviewer 2: This may be because I'm using an old version of sphinx. I think more recent versions default to python highlighting

Reviewer 2: It is now allowed to omit the argument in more recent versions of sphinx
https://github.com/sphinx-doc/sphinx/pull/6015
I just need to update my environment!

        dimensions:
          time=4;
          nv=2;
        variables:
          float temperature(time,lat,lon);
            temperature:long_name="surface air temperature";
            temperature:cell_methods="time: minimum within years time: mean over years";
            temperature:units="K";
          double time(time);
            time:climatology="climatology_bounds";
            time:units="days since 1960-1-1";
          double climatology_bounds(time,nv);
        data:  // time coordinates translated to date/time format
          time="1960-4-16", "1960-7-16", "1960-10-16", "1961-1-16" ;
          climatology_bounds="1960-3-1",  "1990-6-1",
                             "1960-6-1",  "1990-9-1",
                             "1960-9-1",  "1990-12-1",
                             "1960-12-1", "1991-3-1" ;

> 

Author: 👍

  If a climatological time axis is detected when loading NetCDF -
  indicated by the format described above - the ``climatological`` property
  of the Iris coordinate will be set to ``True``.

.. admonition:: New Chunking Strategy

  Iris now makes better choices of Dask chunk sizes when loading from NetCDF
  files: If a file variable has specified, small chunks, Iris will now choose
  Dask chunks which are a multiple of these up to a default target size.

  This is particularly relevant to files with an unlimited dimension, which
  previously could produce a large number of small chunks. This had an adverse
  effect on performance.

  In addition, Iris now takes its default chunksize from the default configured
  in Dask itself, i.e. ``dask.config.get('array.chunk-size')``.

.. admonition:: Lazy Statistics

  Several statistical operations can now be done lazily, taking advantage of the
  performance improvements offered by Dask:

  * :meth:`~iris.cube.Cube.aggregated_by`
  * :class:`~iris.analysis.RMS` (more detail below)
  * :class:`~iris.analysis.MEAN`

> |

Reviewer: @trexfeathers Is this `|` intentional, if so what does it do?

Author: It adds a blank line, making a clearer distinction between the final showcase and the first bullet point:

![Screenshot from 2019-10-04 14-18-12](https://user-images.githubusercontent.com/40734014/66210502-ed849600-e6b1-11e9-95a3-6863eead23fa.png)

Reviewer: @trexfeathers How does it look with a divider instead? i.e., see [reStructuredText transitions](http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#transitions)

* Cube data equality testing (and hence cube equality) now uses a more relaxed
  tolerance : This means that some cubes may now test 'equal' that previously
  did not.
  Previously, Iris compared cube data arrays using "abs(a - b) < 1.e-8".
  We now apply the default operation of :func:`numpy.allclose` instead,

>   which is equivalent to "abs(a - b) < (1.e-8 + 1.e-5 * b)".

Author: Have reformatted

* Added support to render HTML for :class:`iris.cube.CubeList`s in Jupyter Notebooks and JupyterLab.

> * Loading CellMeasures with non-float values is now supported.

Reviewer 3: I think this should be "integer values".  It won't cope with strings !

Author: 👍

> * New coordinate system: :class:`iris.coord_systems.Geostationary`, based on

Reviewer 3: Do we also load+save these in netcdf ?  If so, should say so (like the one below).

Author: 👍

  the `CF Geostationary projection definition <http://cfconventions
  .org/cf-conventions/cf-conventions.html#_geostationary_projection>`_.
* :class:`iris.coord_systems.VerticalPerspective` can now be saved to and
  loaded from NetCDF files.
* :class:`iris.experimental.regrid.PointInCell` moved to
  :class:`iris.analysis.PointInCell` to make this regridding scheme public
* Iris now supports standard name modifiers. See `Appendix C, Standard Name Modifiers <http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#standard-name-modifiers>`_ for more information.
* :meth:`iris.cube.Cube.remove_cell_measure` now also allows removal of a cell
  measure by its name (previously only accepted a CellMeasure object).
* The :data:`iris.analysis.RMS` aggregator now supports a lazy calculation.
  However, the "weights" keyword is not currently supported by this, so a
  *weighted* calculation will still return a realised result, *and* force
  realisation of the original cube data. **NOTE:** lazy calculation with RMS

>   is currently only supported for un-masked arrays.

Author: Having read in more detail I see your point, have remoted the NOTE

Iris 2.3.0 Dependency Updates

> ===========================

Author: Thanks, missed that after I added ".0"

> * The Proj4 pin has been moved from version 5 to version 6 pending `fixes to

Author: 👍

  some cartopy tests <https://github
  .com/SciTools/cartopy/pull/1289#pullrequestreview-272774087>`_.

> * Iris now requires Dask > 1.2 to allow for improved coordinate equality

Author: 👍

  checks.

Bugs Fixed
==========
* Cube equality of boolean data is now handled correctly.

> * Iris now supports NetCDF Climate and Forecast (CF) Metadata Conventions 1.7 (see `CF 1.7 Conventions Document <http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html>`_ for more information)

Author: I've moved this to new features

* Fixed a bug where cell measures were incorrect after a cube
  :meth:`~iris.cube.Cube.transpose` operation.  Previously, this resulted in
  cell-measures that were no longer correctly mapped to the cube dimensions.
* The :class:`~iris.coords.AuxCoord` disregarded masked points and bounds, as did the :class:`~iris.coords.DimCoord`.
  Fix permits an :class:`~iris.coords.AuxCoord` to contain masked points/bounds, and a TypeError exception is now

>   raised when attempting to create or set the points/bounds of a :class:`~iris.coords.DimCoord` with masked arrays.

Author: 👍

* :class:`iris.coord_systems.VerticalPerspective` coordinate system now uses
  the `CF Vertical perspective definition <http://cfconventions
  .org/cf-conventions/cf-conventions.html#vertical-perspective>`_; had been
  erroneously using Geostationary.
* :class:`~iris.coords.CellMethod` will now only use valid `NetCDF name tokens <https://www.unidata.ucar.edu/software/netcdf/docs/netcdf_data_set_components.html#object_name>`_ to reference the coordinates involved in the statistical operation.

> * The following var_names will now only use valid `NetCDF name tokens

Author: 👍

  <https://www.unidata.ucar
  .edu/software/netcdf/docs/netcdf_data_set_components.html#object_name>`_ to
  reference the said NetCDF variable name. Note that names with a leading
  underscore are not permitted.
    - :attr:`iris.aux_factory.AuxCoordFactory.var_name`
    - :attr:`iris.coords.CellMeasure.var_name`
    - :attr:`iris.coords.Coord.var_name`
    - :attr:`iris.coords.AuxCoord.var_name`
    - :attr:`iris.cube.Cube.var_name`
* Cubes are now rendered more robustly by Jupyter.

>   :meth:`~iris.cube.Cube._repr_html_` no longer crashes for a cube with

Author: @stephenworsley and I have rephrased, since it would be inappropriate to reference a private method anyway. (is that why the link wouldn't function?)

  attributes containing ``\n``.
* Variables which reference themselves in ``cell_measures`` can now be read.
* :func:`~iris.plot.quiver` now handles circular coordinates.

> * The name of cubes loaded from abf/abl files has been corrected.

Author: 👍

* Fixed a bug in UM file loading, where any landsea-mask-compressed fields
  (i.e. with LBPACK=x2x) would cause an error later, when realising the data.

> * :meth:`iris.cube.Cube.coll\apsed` now handles partial collapsing of

Author: I will double check all the links next time!

  multidimensional coordinates that have bounds.
* Fixed a bug in the :data:`~iris.analysis.PROPORTION` aggregator, where cube
  data in the form of a masked array with ``array.mask=False`` would cause an
  error, but possibly only later when the values are actually realised.
  ( Note: since netCDF4 version 1.4.0, this is now a common form for data
  loaded from netCDF files ).
* Fixed a bug where plotting a cube with a
  :class:`iris.coord_systems.LambertConformal` coordinate system would result
  in an error.  This would happen if the coordinate system was defined with one
  standard parallel, rather than two.
  In these cases, a call to
  :meth:`~iris.coord_systems.LambertConformal.as_cartopy_crs` would fail.
* :meth:`iris.cube.Cube.aggregated_by` now gives correct values in points and
  bounds when handling multidimensional coordinates.
* Fixed a bug in the :meth:`iris.cube.Cube.collapsed` operation, which caused
  the unexpected realization of any attached auxiliary coordinates that were
  *bounded*.  It now correctly produces a lazy result and does not realise
  the original attached AuxCoords.

Documentation Changes
=====================

> * Added a gallery example showing how to concatenate NEMO ocean model data.

Author: 👍

* Added an example in the

>   `Loading Iris Cubes: Constraining on Time <https://scitools.org.uk/iris/docs/latest/userguide/loading_iris_cubes.html#constraining-on-time>`_

Author: 👍

  Userguide section, demonstrating how to load data within a specified date
  range.
* Added notes to the :func:`iris.load` documentation, and the userguide

>   `Loading Iris Cubes <https://scitools.org.uk/iris/docs/latest/userguide/loading_iris_cubes.html>`_

Author: 👍

What's New in Iris 2.3.0
************************

:Release: 2.3.0
:Date: 2019-10-04

This document explains the new/changed features of Iris in version 2.3.0
(:doc:`View all changes <index>`.)

Iris 2.3.0 Features
===================
.. _showcase:

.. admonition:: Increased Support for CF 1.7

  Several changes have contributed to our increased support for CF 1.7:

  * climatological coordinate support
  * standard name modifier support
  * name tokens enforcement in cell methods
  * coordinate var_name, CF_CONVENTIONS_VERSION advanced to 1.7

>   You can read more about each of these below.

Reviewer 2: I think this showcase needs some changing. We need to get the messaging of this right.
My proposal would be something along the lines of:

```
.. admonition:: Support for CF 1.7

  Several changes have contributed to our increased support for CF conventions
  including some CF 1.7 additions. This has included support for the
  following:

  * Climatological Coordinates
  * Standard name modifiers 
  * The Geostationary coordinate system

  More detail about these additions is outlined below.

  Furthermore, Iris now saves netCDF files with the CF conventions attribute set to 1.7.
```

Reviewer 2: New proposal:
```
.. admonition:: Support for CF 1.7

  We have introduced several changes that contribute to Iris's support for the
  CF Conventions, including some CF 1.7 additions. We are now able to support:  

  * Climatological Coordinates
  * Standard name modifiers 
  * The Geostationary coordinate system

  For more details about these additions, see below.

  Additionally, the conventions attribute, added by Iris when saving to NetDF,
  has been updated to "CF-1.7", accordingly.
```

Author: 👍

    .. code-block:: none

        dimensions:
          time=4;
          nv=2;
        variables:
          float temperature(time,lat,lon);
            temperature:long_name="surface air temperature";
            temperature:cell_methods="time: minimum within years time: mean over years";
            temperature:units="K";
          double time(time);
            time:climatology="climatology_bounds";
            time:units="days since 1960-1-1";
          double climatology_bounds(time,nv);
        data:  // time coordinates translated to date/time format
          time="1960-4-16", "1960-7-16", "1960-10-16", "1961-1-16" ;
          climatology_bounds="1960-3-1",  "1990-6-1",
                             "1960-6-1",  "1990-9-1",
                             "1960-9-1",  "1990-12-1",
                             "1960-12-1", "1991-3-1" ;

  If a climatological time axis is detected when loading NetCDF -
  indicated by the format described above - the ``climatological`` property
  of the Iris coordinate will be set to ``True``.

.. admonition:: New Chunking Strategy

  Iris now makes better choices of Dask chunk sizes when loading from NetCDF

>   files: If a file variable has specified, small chunks, Iris will now choose

Reviewer 2: I remember this phrasing tripping me up in the PR it was included. I think the ordering of the adjectives are wrong.
I think it should be:
`If a file variable has small, specified chunks, Iris will now choose`

Author: 👍

  Dask chunks which are a multiple of these up to a default target size.

  This is particularly relevant to files with an unlimited dimension, which
  previously could produce a large number of small chunks. This had an adverse
  effect on performance.

  In addition, Iris now takes its default chunksize from the default configured
  in Dask itself, i.e. ``dask.config.get('array.chunk-size')``.

.. admonition:: Lazy Statistics

  Several statistical operations can now be done lazily, taking advantage of the
  performance improvements offered by Dask:

  * :meth:`~iris.cube.Cube.aggregated_by`
  * :class:`~iris.analysis.RMS` (more detail below)
  * :class:`~iris.analysis.MEAN`

----

* Cube data equality testing (and hence cube equality) now uses a more relaxed
  tolerance : This means that some cubes may now test 'equal' that previously
  did not.
  Previously, Iris compared cube data arrays using "abs(a - b) < 1.e-8".
  We now apply the default operation of :func:`numpy.allclose` instead,
  which is equivalent to "abs(a - b) < (1.e-8 + 1.e-5 * b)".

> * Added support to render HTML for :class:`iris.cube.CubeList`s in Jupyter Notebooks and JupyterLab.

Reviewer 2: This may be due to my old version of Sphinx, but it is complaining about this line and in the built copy it doesn't correctly create the link to iris.cube.CubeList

You could also argue that we should use the `~` i.e. :class:`~iris.cube.CubeList`so that it reads more nicely:
`Added support to render HTML for CubeList in Jupyter Notebooks and JupyterLab`
where cube list would link off to the ref for CubeList

Reviewer 3: I caught this one.  I'm pretty sure the cause is the "s" immediately following.
I know there is a way to fix this..

In the 2.2 whatsnew I find : `:class:\`iris.coords.AuxCoord\`'s can now test as 'equal'`
That at least seems to work.

Author: Went for @lbdreyer's idea

* Loading CellMeasures with non-float values is now supported.

> * New coordinate system: :class:`iris.coord_systems.Geostationary`, based on

Reviewer 2: This should be in the showcase as Geostationary was introduced in CF1.7

Author: 👍

* :class:`iris.coord_systems.VerticalPerspective` coordinate system now uses
  the `CF Vertical perspective definition <http://cfconventions
  .org/cf-conventions/cf-conventions.html#vertical-perspective>`_; had been
  erroneously using Geostationary.
* :class:`~iris.coords.CellMethod` will now only use valid `NetCDF name tokens <https://www.unidata.ucar.edu/software/netcdf/docs/netcdf_data_set_components.html#object_name>`_ to reference the coordinates involved in the statistical operation.
* The following var_names will now only use valid `NetCDF name tokens
  <https://www.unidata.ucar
  .edu/software/netcdf/docs/netcdf_data_set_components.html#object_name>`_ to
  reference the said NetCDF variable name. Note that names with a leading
  underscore are not permitted.
    - :attr:`iris.aux_factory.AuxCoordFactory.var_name`
    - :attr:`iris.coords.CellMeasure.var_name`
    - :attr:`iris.coords.Coord.var_name`
    - :attr:`iris.coords.AuxCoord.var_name`
    - :attr:`iris.cube.Cube.var_name`
* Cubes are now rendered more robustly by Jupyter.
  :meth:`~iris.cube.Cube._repr_html_` no longer crashes for a cube with
  attributes containing ``\n``.

> * Variables which reference themselves in ``cell_measures`` can now be read.

Reviewer 2: This is a bit unclear.
I would specify that the you are talking about netCDF here.
Something like:
```suggestion
* NetCDF Variables which reference themselves in ``cell_measures`` can now be read.
```

Author: 👍
